kernel/queue.c

```diff
@@ -205,21 +203,25 @@ static void *k_queue_poll(struct k_queue *queue, s32_t timeout)
 	k_poll_event_init(&event, K_POLL_TYPE_FIFO_DATA_AVAILABLE,
 			  K_POLL_MODE_NOTIFY_ONLY, queue);
 
-	event.state = K_POLL_STATE_NOT_READY;
+	do {
+		event.state = K_POLL_STATE_NOT_READY;
 
-	err = k_poll(&event, 1, timeout);
-	if (err) {
-		return NULL;
-	}
+		err = k_poll(&event, 1, timeout);
+		if (err) {
```

Comment:
Are you missing the check for -EINVAL?

Reply:
Does it matter? if it's any error you return NULL anyway?

Reply:
ah ok.

Reply:
Yep, any error here will cause it to return NULL which is what the user of k_queue_cancel_wait would expect according to its documentation.